### ISSUE TITLE
Add toolkit methods that manage memory for char * return types

### ISSUE THREAD

**Issue by michaeltryby** [OpenWaterAnalytics/EPANET#309]
To make the Toolkit API easier to use methods that return char * types should allocate memory. The caller is responsible for freeing memory. 

This helps with conversion to variable length ID names and also facilitates wrapping of the functions in SWIG.

**Comment by LRossman** [comment]
I'm sorry @michaeltryby , but I am totally against having the API allocate memory on its own which it doesn't release. This seems like a recipe for disaster (ok, I'm overstating things for effect). It's also considered [bad programming practice](https://stackoverflow.com/questions/13625388/is-it-bad-practice-to-allocate-memory-in-a-dll-and-give-a-pointer-to-it-to-a-cli) by some. Unless one is using smart pointers (which C doesn't have) one should stick to the rule that the function or module that allocates the memory should be responsible for releasing it.

**Comment by jamesuber** [comment]
smart pointers are wonderful. I remember when @samhatchett introduced them to me years ago. Another reason for C++, I guess (or boost?)

Sorry, i know that wasn’t exactly helpful to the task at hand.

**Comment by michaeltryby** [comment]
@LRossman I understand you come from the conservative school of memory management. And conservative is good policy when we are talking memory. But I think you are wrong on this one. I tried several different strategies to manage memory in the EPANET and SWMM output APIs and this is by far the best. I was able to write a compact API that returns a lot of data to the user in every single call with no hassles. All for the price of remembering to free the memory. It was an easy trade off to make. 

Furthermore, this approach has been in production going on a year now and we haven’t experienced any “disasters”. It gets used every time someone creates a PR in the OWA EPANET and SWMM repos. And @bemcdonnell makes intensive use of it in EmNet’s workflow. 

On the python side this approach allows the wrappers to hand off memory between C and python automatically. This is de rigueur since python programmers take memory for granted. No problems. Python programmers are happy. 

Lastly, what’s the worst that can happen, a catastrophic memory leak? Did you know that when you build the unit tests in debug mode it automatically flags leaks for you? Imagine that! Unit testing makes the leaks easy to find. 

So this is how we are going to do it. And no, we aren’t courting “disaster.”

**Comment by LRossman** [comment]
Wow. After devoting a good chunk of my professional career to creating and maintaining EPANET and now volunteering in my retirement to add some much needed computational upgrades to it, I'm told "So this is how we are going to do it". Pretty sad. So much for the consensus approach to building an open source project.

**Comment by michaeltryby** [comment]
Well I guess it takes one to know one, cause you never struck me as a consensus builder either. 

I think if you are honest with yourself, you’ll realize that you have  programmed EPANET and SWMM your way for your entire career. 

Perhaps it’s time to give other people a chance to do it their way for a change.

**Comment by eladsal** [comment]
"So this is how we are going to do it" - who is "we"?

@michaeltryby any chance your account was hacked?

**Comment by samhatchett** [comment]
Hi, just catching up here. First of all - WOW, @michaeltryby you are being rude.

Second of all, I agree on technical grounds with @LRossman about memory management. Especially if this is liked in a C++ project: if the library allocates memory using one version of the stdlib, and the client code tries to free it using a different standard library, then trouble will ensue. It's bad form. I believe normal practice is to provide a free function to ensure consistency.

**Comment by michaeltryby** [comment]
Sorry @LRossman frustration got the better of me 😔

**Comment by LRossman** [comment]
I say forgive and forget.

**Comment by samhatchett** [comment]
Centering back on the technical discussion, if need be I can open a new issue...

I will direct your attention to the wonderfully-written MSDN article [Allocating and freeing memory across module boundaries](https://blogs.msdn.microsoft.com/oldnewthing/20060915-04/?p=29723) - wherein this very issue is discussed. 

Regardless of whether this works in testing, expecting the client to free memory is trouble. If we are producing and distributing a compiled library, we won't have control over the deallocator that gets called on library-allocated memory. It's unsafe and against common practice.

**Comment by LRossman** [comment]
@samhatchett the error manager that was added to the 2.2 dev code some months back already uses this unsafe approach to return an error message to `EN_checkerror`. Please refer to #304, which I just closed today. You might want to re-open it if you feel it's still problematic.

**Comment by michaeltryby** [comment]
@samhatchett nice article thanks for sending it. It clearly describes the problem. 

So if I understand the article correctly, it the dll exports a `free()` function such as `EN_free()` when it gets used the same C/C++ runtime (the dll's) allocates and frees the memory. If it doesn't get used then a memory leak or worse occurs. Is that correct.

**Comment by samhatchett** [comment]
I think that's right. This concept is also known as "ownership semantics" and is simplified by the C++ auto/shared pointers. In C, however, we just need to make sure that the freeing function is under our control.

**Comment by michaeltryby** [comment]
Yes that is my understanding as well. My thought was to go ahead and expose helper functions for freeing memory in our API's, configure the SWIG wrappers to use them, and beef up the documentation. 

@LRossman How does that sound? I know it still leaves open a possible problem if an unsuspecting developer (such as myself) fails to use the API to free memory. But I think the advantages on the python side outweigh the risks in this instance.

**Comment by LRossman** [comment]
Responding to @michaeltryby suggestion to add an `EN_free()` function to the API:
Are we still talking about modifying every API function that returns a character string or just modifying EN_checkerror. If the former, then every time that one of the `EN_getxxxid` functions gets called the API user has to call `EN_free` once they are done using the returned ID and before they call the ID getter again using the same character pointer as an argument. Seems like a lot of bending over backwards just to make life simpler for a Python wrapper.

**Comment by michaeltryby** [comment]
@LRossman yes every function call that allocates memory will have to call EN_free() afterwards. That will be part of the new contract. Another potential solution is to version the API. 

Python wrappers are a new feature. I think they will prove to be well worth the trouble.
